### PR TITLE
Selfie: Take aspect ratio into consideration when drawing grid

### LINKED ISSUE #41: 4:3 grid is drawn outside of preview image

### PR DESCRIPTION
**Pull request by luk1337**
Change-Id: Id889c5fcf601e6a9756a668963e8c9d309e58d28

### CODE REVIEW

**Comment by SebaUbuntu** [comment]
On 4:3 grid is slightly too high, we could just set height and width since we have them

**Comment by luca020400** [comment]
This looks horrible and beautiful at the same time 
Nice job

**Comment by SebaUbuntu** [comment]
Screenshot_20220811-161050
Handles this strange case where 16:9 doesn't fit horizontally

File: app/src/main/java/org/lineageos/selfie/MainActivity.kt [hunk 1]
```diff
@@ -413,6 +423,9 @@ class MainActivity : AppCompatActivity() {
     @androidx.camera.core.ExperimentalZeroShutterLag
     @androidx.camera.view.video.ExperimentalVideo
     private fun bindCameraUseCases() {
+        // Unbind previous use cases
+        cameraController.unbind()
```
Comment (luca020400) [inline app/src/main/java/org/lineageos/selfie/MainActivity.kt:438]: This also seems to fix my issues with landscape photos?

File: app/src/main/java/org/lineageos/selfie/MainActivity.kt [hunk 2]
```diff
@@ -556,6 +572,18 @@ class MainActivity : AppCompatActivity() {
...
+    /**
+     * Update grid aspect ratio based on current preview size
+     */
+    @androidx.camera.view.TransformExperimental
+    private fun updateGridAspectRatio() {
+        val outputTransform = viewFinder.outputTransform ?: return
+        val size = outputTransform.getViewPortSizeUnrestricted()
+
+        // For somehow reason, height and width are inverted :D
+        gridView.previewSize = Size(size.height, size.width)
```
Comment (luca020400) [inline app/src/main/java/org/lineageos/selfie/MainActivity.kt]: It's likely the case it uses landscape

Comment (SebaUbuntu) [reply app/src/main/java/org/lineageos/selfie/MainActivity.kt]: It's inverted on all the devices I've tested with

File: app/src/main/java/org/lineageos/selfie/BitmapExt.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                px++
+            } else {
+                break
+            }
+        }
+
+        return width - px
+    }
+
+internal val Bitmap.nonTransparentSize: Size
```
Comment (luca020400) [inline app/src/main/java/org/lineageos/selfie/BitmapExt.kt]: That's, wow
Better than before, but wow

File: app/src/main/java/org/lineageos/selfie/BitmapExt.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+internal val Bitmap.nonTransparentHeight: Int
+    get() {
+        val width = width
+        val height = height
+
+        var px = 0
+
+        // Count transparent pixels from top
+        for (i in 0 until height - 1) {
+            if (getPixel(width / 2, i) == 0) {
```
Comment (luca020400) [inline app/src/main/java/org/lineageos/selfie/BitmapExt.kt]: Why from half screen?
Isn't it the same from X set to 0?

File: app/src/main/java/org/lineageos/selfie/BitmapExt.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+internal val Bitmap.nonTransparentWidth: Int
+    get() {
+        val width = width
+        val height = height
+
+        var px = 0
+
+        // Count transparent pixels from left
+        for (i in 0 until width - 1) {
+            if (getPixel(i, height / 2) == 0) {
```
Comment (luca020400) [inline app/src/main/java/org/lineageos/selfie/BitmapExt.kt]: Ditto

File: app/src/main/java/org/lineageos/selfie/BitmapExt.kt (hunk 4 above)
Comment (luca020400) [reply app/src/main/java/org/lineageos/selfie/BitmapExt.kt]: Never mind, I hit send review and it hit me.

File: app/src/main/java/org/lineageos/selfie/ui/GridView.kt [hunk 6]
```diff
@@ -43,12 +49,15 @@ class GridView(context: Context?, attributeSet: AttributeSet?) : View(context, a
             GridMode.ON_GOLDENRATIO -> 3
         }
 
-        if (size <= 0) {
+        if (size <= 0 || previewSize == null) {
```
Comment (SebaUbuntu) [inline app/src/main/java/org/lineageos/selfie/ui/GridView.kt]: `val previewSize = previewSize ?: return`

Comment (luk1337) [reply app/src/main/java/org/lineageos/selfie/ui/GridView.kt]: not sure if that's worth it...

Comment (SebaUbuntu) [reply app/src/main/java/org/lineageos/selfie/ui/GridView.kt]: This way you can get rid of height and width variables and !! operators

**Pull request merged**